kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ConsumerOffsetsDaoImpl.java

```diff
@@ -0,0 +1,234 @@
+      long latestOffset = getOffset(latestOffsets, tp);
+      if (currentOffset < 0 || latestOffset < 0) {
+        // log.debug("invalid offsets for topic={} consumerId={} current={} latest={}",
+        //     tp.topic(),
+        //     consumerId,
+        //     currentOffset,
+        //     latestOffset);
+        continue;
+      }
+      cgOffsets.addOffset(
+          tp.topic(),
+          consumerId,
+          clientId,
+          tp.partition(),
+          currentOffset,
+          latestOffset
+      );
+    }
+    // ahu todo: instanceId
+    return cgOffsets;
+  }
+
+  public ConsumerGroupLag getConsumerGroupOffsets(
+      String clusterId,
+      String consumerGroupId,
+      IsolationLevel isolationLevel
+  ) throws InterruptedException, ExecutionException, TimeoutException {
+    final ConsumerGroupDescription cgDesc = getConsumerGroupDescription(consumerGroupId);
+    return getConsumerGroupOffsets(cgDesc, isolationLevel).setClusterId(clusterId).build();
+  }
+
+  private ConsumerGroupLag.Builder getConsumerGroupOffsets(
+      ConsumerGroupDescription desc,
+      IsolationLevel isolationLevel
+  ) throws InterruptedException, ExecutionException, TimeoutException {
+    final Map<TopicPartition, OffsetAndMetadata> fetchedCurrentOffsets =
+        getCurrentOffsets(desc.groupId());
+
+    ListOffsetsOptions listOffsetsOptions = new ListOffsetsOptions(isolationLevel)
+        .timeoutMs(consumerMetadataTimeout);
+
+    Map<TopicPartition, OffsetSpec> latestOffsetSpecs = fetchedCurrentOffsets.keySet().stream()
+        .collect(Collectors.toMap(Function.identity(), tp -> OffsetSpec.latest()));
+
+    ListOffsetsResult latestOffsetResult = kafkaAdminClient.listOffsets(
+        latestOffsetSpecs, listOffsetsOptions);
+    Map<TopicPartition, ListOffsetsResultInfo> latestOffsets = latestOffsetResult.all()
+        .get(consumerMetadataTimeout, TimeUnit.MILLISECONDS);
+
+    return getConsumerGroupOffsets(
+        desc,
+        fetchedCurrentOffsets,
+        latestOffsets
+    );
+  }
+
+  public Set<String> getConsumerGroups()
+      throws InterruptedException, ExecutionException, TimeoutException {
+    return Sets.newLinkedHashSet(
+        Iterables.transform(kafkaAdminClient
+                .listConsumerGroups(new ListConsumerGroupsOptions()
+                    .timeoutMs(consumerMetadataTimeout))
+                .all()
+                .get(consumerMetadataTimeout, TimeUnit.MILLISECONDS),
+            ConsumerGroupListing::groupId));
+  }
+
+  public ConsumerGroupDescription getConsumerGroupDescription(
+      String consumerGroupId
+  ) throws InterruptedException, ExecutionException, TimeoutException {
+    Map<String, ConsumerGroupDescription> allCgDesc =
+        getAllConsumerGroupDescriptions(ImmutableSet.of(consumerGroupId));
+    return allCgDesc.get(consumerGroupId);
+  }
+
+  public Map<String, ConsumerGroupDescription> getAllConsumerGroupDescriptions(
```

Comment:
Which API does this satisfy? It doesn't seem too useful to get detailed member information for multiple groups at the same time.

Reply:
It's currently only used to grab the description of single consumer groups (called by `getConsumerGroupDescription`). I was thinking of leaving it as is since it seems blueway uses this for c3/cloud UI's lag display (which the rest proxy needs to support in the future)?